Automatically get Ganon Boss Key when meeting goal. (No longer tied to LACS.)

Cuphat opened the pull request:
Demonstration: https://www.youtube.com/watch?v=rPEXP52sYbw

This adds new options for Ganon Boss Key: Medallions, Stones, Dungeons, and Tokens. When reaching the goal amount of the type selected, you are automatically rewarded the Ganon BK. This allows LACS to be returned to a normal check but still having the flexibility of the bridge and beating Ganon not being tied together. Ganon BK can still be placed on LACS if wanted.

The LACS condition settings have now been hidden from the GUI. They still exist, however, and can be utilized on the command line or in a plando.

Cuphat committed:
Functional-esque start of auto Ganon BK.

Cuphat committed:
Settings, Python, and Logic for automatic Ganon BK.

Cuphat committed:
Remove LACS Settings from GUI

These settings still exist and can be utilized in plando, but they are no longer displayed in the GUI. Presets also updated to take this into account.

Zannick commented on ItemPool.py:
```diff
@@ -1314,6 +1314,12 @@ def get_pool_core(world):
         placed_items['Ganons Tower Boss Key Chest'] = 'Boss Key (Ganons Castle)'
 
+    if world.settings.shuffle_ganon_bosskey in ['stones', 'medallions', 'dungeons', 'tokens']:
+         placed_items['Gift from Sages'] = 'Boss Key (Ganons Castle)'
```
nit: too many spaces on that indent

Cuphat replied:
Easy enough to fix.

Zannick commented on SettingsList.py:
```diff
@@ -3140,8 +3144,26 @@ def __init__(self, name, gui_text, min, max, default, step=1,
+            when reaching the target number of Medallions.
+                        
+            'Dungeons': Ganon's Castle Boss Key will be awarded
+            when reaching the target number of Dungeon Rewards.
```
This includes the 1 in Link's Pocket, right? Should that be noted here? (Should the minimum be 2?)

Cuphat replied:
It does include the 1 in Link's Pocket, same as LACS Condition and Bridge Condition. Minimum of 2 would make sense, but it'd also make sense in those as well.

Cuphat commented on ASM/c/ganon_boss_key.c:
```diff
@@ -0,0 +1,52 @@
+                        1, 2, 2, 3, 2, 3, 3, 4 };
+
+// https://www.geeksforgeeks.org/count-set-bits-in-an-integer/
+unsigned int countSetBitsRec(unsigned int num)
```
`__builtin_popcount` seems to break armips.
```build.asm(104) error: Undefined external symbol __popcountsi2 in file bundle.o
build.asm(104) error: Undefined external symbol __popcountsi2 in file bundle.o
build.asm(104) error: Undefined external symbol __popcountsi2 in file bundle.o
build.asm(119) warning: Content of areas 80400000 and 80400034 overlap
Aborting.```

Cuphat committed:
Triforce Hunt disables extra Ganon's Boss Key settings.

Cuphat committed:
Update 'item.goalitem' for Ganon BK settings.

LACS Condition is no longer considered for considering an item a goal item unless Ganon's BK is placed there.

Cuphat commented on ASM/c/ganon_boss_key.c:
```diff
@@ -0,0 +1,52 @@
+        }
+
+        if (give_boss_key) {
+            push_delayed_item(0x03);
```
That's doable, but it'd also mean that getting the boss key and doing a Don't Save/Don't Continue would mean that you no longer would get the boss key until you actually reset the game. Maybe we should do some further testing to see if the boss key being given multiple times is something that could happen?
